File: project/backend/app/models/medical_record.py
"""
Medical record model for storing consultation notes and prescriptions with versioning
"""
from sqlalchemy import Column, Integer, Text, ForeignKey
from sqlalchemy.orm import relationship

from app.db.base import BaseModel


class MedicalRecord(BaseModel):
    """
    Medical record model for storing consultation notes, diagnoses, and prescriptions.
    Supports version control for record updates.
    
    Attributes:
        id: Primary key (inherited from Base)
        patient_id: Foreign key to patients table
        doctor_id: Foreign key to doctors table
        appointment_id: Foreign key to appointments table (nullable)
        consultation_notes: Doctor's consultation notes
        diagnosis: Medical diagnosis
        prescription: Prescription details (medication, dosage, frequency, duration)
        version_number: Version number for record versioning
        parent_record_id: Foreign key to parent record for version tracking
        created_by: Foreign key to users table (who created this record)
        created_at: Timestamp when record was created (inherited from Base)
    
    Relationships:
        patient: Associated patient
        doctor: Associated doctor
        appointment: Associated appointment
        creator: User who created this record
        parent_record: Parent record for version tracking
        child_records: Child records (newer versions)
    """
    __tablename__ = "medical_records"
    
    patient_id = Column(Integer, ForeignKey("patients.id", ondelete="CASCADE"), nullable=False, index=True)
    doctor_id = Column(Integer, ForeignKey("doctors.id", ondelete="CASCADE"), nullable=False, index=True)
    appointment_id = Column(Integer, ForeignKey("appointments.id", ondelete="SET NULL"), index=True)
    consultation_notes = Column(Text)
    diagnosis = Column(Text)
    prescription = Column(Text)
    version_number = Column(Integer, default=1, nullable=False)
    parent_record_id = Column(Integer, ForeignKey("medical_records.id", ondelete="SET NULL"), index=True)
    created_by = Column(Integer, ForeignKey("users.id", ondelete="SET NULL"))
    
    # Relationships
    patient = relationship("Patient", back_populates="medical_records")
    doctor = relationship("Doctor", back_populates="medical_records")
    appointment = relationship("Appointment", back_populates="medical_records")
    creator = relationship("User", foreign_keys=[created_by], back_populates="medical_records_created")
    parent_record = relationship("MedicalRecord", remote_side="MedicalRecord.id", foreign_keys=[parent_record_id], backref="child_records")
    
    def __repr__(self):
        return f"<MedicalRecord(id={self.id}, patient_id={self.patient_id}, version={self.version_number})>"

File: project/mobile/lib/models/medical_record.dart
/// Medical record model representing consultation notes and prescriptions
class MedicalRecord {
  final int id;
  final int patientId;
  final int doctorId;
  final int? appointmentId;
  final String? consultationNotes;
  final String? diagnosis;
  final String? prescription;
  final int versionNumber;
  final int? parentRecordId;
  final int createdBy;
  final String? doctorName;
  final DateTime createdAt;
  final bool? isTampered;

  MedicalRecord({
    required this.id,
    required this.patientId,
    required this.doctorId,
    this.appointmentId,
    this.consultationNotes,
    this.diagnosis,
    this.prescription,
    required this.versionNumber,
    this.parentRecordId,
    required this.createdBy,
    this.doctorName,
    required this.createdAt,
    this.isTampered,
  });

  factory MedicalRecord.fromJson(Map<String, dynamic> json) {
    return MedicalRecord(
      id: json['id'] as int,
      patientId: json['patient_id'] as int,
      doctorId: json['doctor_id'] as int,
      appointmentId: json['appointment_id'] as int?,
      consultationNotes: json['consultation_notes'] as String?,
      diagnosis: json['diagnosis'] as String?,
      prescription: json['prescription'] as String?,
      versionNumber: json['version_number'] as int? ?? 1,
      parentRecordId: json['parent_record_id'] as int?,
      createdBy: json['created_by'] as int,
      doctorName: json['doctor_name'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      isTampered: json['is_tampered'] as bool?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'patient_id': patientId,
      'doctor_id': doctorId,
      'appointment_id': appointmentId,
      'consultation_notes': consultationNotes,
      'diagnosis': diagnosis,
      'prescription': prescription,
      'version_number': versionNumber,
      'parent_record_id': parentRecordId,
      'created_by': createdBy,
      'doctor_name': doctorName,
      'created_at': createdAt.toIso8601String(),
      'is_tampered': isTampered,
    };
  }

  bool get hasConsultationNotes => consultationNotes != null && consultationNotes!.isNotEmpty;
  bool get hasPrescription => prescription != null && prescription!.isNotEmpty;
  bool get isOriginal => versionNumber == 1;
}
